src/dimension_slice.c

```diff
@@ -1179,54 +1224,114 @@ ts_osm_chunk_range_overlaps(int32 osm_dimension_slice_id, int32 dimension_id, in
+	ts_scan_iterator_scan_key_init(&iterator,
+								   Anum_dimension_slice_id,
+								   BTEqualStrategyNumber,
+								   F_INT4EQ,
+								   Int32GetDatum(dimension_slice_id));
 
-	values[AttrNumberGetAttrOffset(Anum_dimension_slice_range_end)] = Int64GetDatum(fd->range_end);
-	doReplace[AttrNumberGetAttrOffset(Anum_dimension_slice_range_end)] = true;
+	ts_scanner_foreach(&iterator)
+	{
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&iterator);
+		if (ti->lockresult != TM_Ok)
+		{
+			if (IsolationUsesXactSnapshot())
+			{
+				/* For Repeatable Read and Serializable isolation level report error
+				 * if we cannot lock the tuple
+				 */
+				ereport(ERROR,
+						(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
+						 errmsg("could not serialize access due to concurrent update")));
+			}
+			else
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_INTERNAL_ERROR),
+						 errmsg("unable to lock hypertable catalog tuple, lock result is %d for "
+								"hypertable "
+								"ID (%d)",
+								ti->lockresult,
+								dimension_slice_id)));
+			}
+		}
+		dimension_slice_formdata_fill(form, ti);
+		ItemPointer result_tid = ts_scanner_get_tuple_tid(ti);
+		tid->ip_blkid = result_tid->ip_blkid;
+		tid->ip_posid = result_tid->ip_posid;
+		success = true;
+		break;
+	}
+	ts_scan_iterator_close(&iterator);
+	return success;
+}
 
-	HeapTuple new_tuple =
-		heap_modify_tuple(tuple, ts_scanner_get_tupledesc(ti), values, isnull, doReplace);
+/* update the tuple at this tid. The assumption is that we already hold a
+ * tuple exclusive lock and no other transaction can modify this tuple
+ * The sequence of operations for any update is:
+ * lock the tuple using lock_hypertable_tuple.
+ * then update the required fields
+ * call dimension_slice_update_catalog_tuple to complete the update.
+ * This ensures correct tuple locking and tuple updates in the presence of
+ * concurrent transactions. Failure to follow this results in catalog corruption
+ */
+static void
+dimension_slice_update_catalog_tuple(ItemPointer tid, FormData_dimension_slice *update)
+{
+	HeapTuple new_tuple;
+	CatalogSecurityContext sec_ctx;
+	Catalog *catalog = ts_catalog_get();
+	Oid table = catalog_get_table_id(catalog, DIMENSION_SLICE);
+	Relation dimension_slice_rel = relation_open(table, RowExclusiveLock);
 
-	ts_catalog_update(ti->scanrel, new_tuple);
+	new_tuple = dimension_slice_formdata_make_tuple(update, dimension_slice_rel->rd_att);
 
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	ts_catalog_update_tid(dimension_slice_rel, tid, new_tuple);
+	ts_catalog_restore_user(&sec_ctx);
 	heap_freetuple(new_tuple);
-	if (should_free)
-		heap_freetuple(tuple);
-
-	return SCAN_DONE;
+	relation_close(dimension_slice_rel, NoLock);
 }
 
 int
-ts_dimension_slice_update_by_id(int32 dimension_slice_id, FormData_dimension_slice *fd_slice)
+ts_dimension_slice_range_update(DimensionSlice *slice)
```

Comment:
This change looks good.
I believe we also need to handle the delete_by_id case. Any tuple change should use this new code path. I think we have a mix due to this API: ts_dimension_slice_scan_by_id_and_lock

If I am reading this correctly, we should use that function (ts_dimension_slice_scan_by_id_and_lock) only for read-only i.e. SELECTS and not for any catalog updates or deletes.